plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/DeployConsolePageParticipant.java

```diff
@@ -0,0 +1,129 @@
+package com.google.cloud.tools.eclipse.appengine.deploy.ui;
+
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import org.eclipse.core.runtime.jobs.IJobChangeEvent;
+import org.eclipse.core.runtime.jobs.Job;
+import org.eclipse.core.runtime.jobs.JobChangeAdapter;
+import org.eclipse.jface.action.Action;
+import org.eclipse.jface.action.IToolBarManager;
+import org.eclipse.jface.resource.ImageDescriptor;
+import org.eclipse.ui.IActionBars;
+import org.eclipse.ui.ISharedImages;
+import org.eclipse.ui.PlatformUI;
+import org.eclipse.ui.console.ConsolePlugin;
+import org.eclipse.ui.console.IConsole;
+import org.eclipse.ui.console.IConsoleConstants;
+import org.eclipse.ui.console.IConsolePageParticipant;
+import org.eclipse.ui.part.IPageBookViewPage;
+
+import com.google.cloud.tools.eclipse.appengine.deploy.standard.StandardDeployJob;
+import com.google.cloud.tools.eclipse.ui.util.Messages;
+
+public class DeployConsolePageParticipant implements IConsolePageParticipant {
+
+  private static final Logger logger = Logger.getLogger(DeployConsolePageParticipant.class.getName());
+
+  private DeployConsole console;
+  private Action terminateAction;
+  private Action closeAction;
+
+  @Override
+  public void init(IPageBookViewPage page, IConsole console) {
+    if (console instanceof DeployConsole) {
+      this.console = (DeployConsole) console;
+    } else {
+      logger.log(Level.SEVERE,
+                 "console is instance of {0}, expected was {1}",
+                 new Object[]{ console.getClass().getName(), DeployConsole.class.getName() });
+    }
+    IActionBars actionBars = page.getSite().getActionBars();
+    configureToolBar(actionBars.getToolBarManager());
+    addJobchangeListener();
+    update();
+  }
+
+  private void configureToolBar(IToolBarManager toolbarManager) {
+    terminateAction = createTerminateAction();
+    toolbarManager.appendToGroup(IConsoleConstants.LAUNCH_GROUP, terminateAction);
+
+    closeAction = createCloseAction();
+    toolbarManager.appendToGroup(IConsoleConstants.LAUNCH_GROUP, closeAction);
+  }
+
+  private void addJobchangeListener() {
+    console.getJob().addJobChangeListener(new JobChangeAdapter() {
```

Comment:
If order of `setJob` and _add-participants_ can vary you'll need to guard against null here.